drivers/meter/ade7880/ade7880.c

```diff
@@ -0,0 +1,487 @@
+	uint32_t temp;
+	/* i rms phase register addr */
+	uint16_t irms_reg;
+	/* v rms phase register addr */
+	uint16_t vrms_reg;
+	/* intermediate values used for computation */
+	uint64_t val_int;
+
+	if (!dev)
+		return -ENODEV;
+
+	/* select the phase for values read */
+	switch (phase) {
+	case ADE7880_PHASE_A:
+		irms_reg = ADE7880_REG_AIRMS;
+		vrms_reg = ADE7880_REG_AVRMS;
+		break;
+	case ADE7880_PHASE_B:
+		irms_reg = ADE7880_REG_BIRMS;
+		vrms_reg = ADE7880_REG_BVRMS;
+		break;
+	case ADE7880_PHASE_C:
+		irms_reg = ADE7880_REG_CIRMS;
+		vrms_reg = ADE7880_REG_CVRMS;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = ade7880_read(dev, irms_reg, &temp);
+	if (ret)
+		return ret;
+
+	// Value in mA
+	dev->irms_val = (uint32_t)(uint64_t)(((uint64_t) temp *
+					      (uint64_t) ADE7880_FS_VOLTAGE
+					      * (uint64_t) ADE7880_CURRENT_TR_FCN) / ((uint64_t)ADE7880_RMS_FS_CODES * 10));
+
+	ret = ade7880_read(dev, vrms_reg, &temp);
+	if (ret)
+		return ret;
+
+	// Value in mV
+	dev->vrms_val = (uint32_t)(uint64_t)(((uint64_t) temp *
+					      (uint64_t) ADE7880_FS_VOLTAGE
+					      * (uint64_t) ADE7880_VOLTAGE_TR_FCN) / ((uint64_t)ADE7880_RMS_FS_CODES * 10));
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ade7880_init(struct ade7880_dev **device,
+		 struct ade7880_init_param init_param)
+{
+	int ret;
+	/* device structure */
+	struct ade7880_dev *dev;
+	/* chip id read value */
+	uint32_t chip_id;
+	/* part id read value */
+	uint32_t part_id;
+
+	dev = (struct ade7880_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* Set power mode */
+	dev->power_mode = init_param.power_mode;
+	if (!init_param.psm0_desc)
+		goto error_dev;
+	dev->psm0_desc = init_param.psm0_desc;
+	if (!init_param.psm1_desc)
+		goto error_dev;
+	dev->psm1_desc = init_param.psm1_desc;
+
+	ret = ade7880_set_power_mode(dev);
+	if (ret)
+		goto error_dev;
+
+	/* Hard reset the device */
+	if (!init_param.reset_desc)
+		goto error_dev;
+	dev->reset_desc = init_param.reset_desc;
+	ret = no_os_gpio_set_value(dev->reset_desc,
+				   NO_OS_GPIO_LOW);
+	if (ret)
+		return ret;
+	// delay reset
+	no_os_mdelay(1);
+	ret = no_os_gpio_set_value(dev->reset_desc,
+				   NO_OS_GPIO_HIGH);
+	if (ret)
+		return ret;
+
+	// wait for device to initialize after hardware reset
+	// >= 100 ms see datasheet.
+	no_os_mdelay(ADE7880_RESET_RECOVER);
+
+	// Select the SPI as the serial interface
+	// Toggle the SS pin 3 times -> see datasheet.
+	if (!init_param.ss_desc)
+		goto error_dev;
+	dev->ss_desc = init_param.ss_desc;
+	for (int i = 0; i < 3; i++) {
+		ret = no_os_gpio_set_value(dev->ss_desc,
+					   NO_OS_GPIO_LOW);
+		no_os_mdelay(1);
+		ret = no_os_gpio_set_value(dev->ss_desc,
+					   NO_OS_GPIO_HIGH);
+		no_os_mdelay(1);
+	}
```

Comment:
The datasheet specifies two ways of selecting the SPI as MCU interface:

```
First, use the SS/HSA pin of the master device (that is, the
microcontroller) as a regular I/O pin and toggle it three times.
Second, execute three SPI write operations to a location in the
address space that is not allocated to a specific ADE7880 register
(for example 0xEBFF, where eight bit writes can be executed).
These writes allow the SS/HSA pin to toggle three times.
```
As far as I understand, you could also issue 3 SPI transactions instead of controlling the SS pin as a GPIO, which would simplify things a bit, since the user doesn't need configure SS a GPIO prior to calling `ade7880_init()`.

Reply:
True. I changed the SPI select method to the one using 3 consecutive write operations.